### ISSUE TITLE
Support Qwen4-Exp layer streaming with SSD-backed PLE N-gram embeddings

### ISSUE THREAD

**Issue by Amix29** [MakazhanAlpamys/Soup#602]
## Problem

PR #572 deliberately keeps `qwen4_exp` / `qwen4_exp_text` out of layer
streaming until architecture parity is demonstrated. That guard is correct,
but it leaves Qwen3.8-Flash-Next without a practical low-memory training path.

Qwen4-Exp places its PLE N-gram embedding inside decoder layer 2. The public
Qwen3.8-Flash-Next model card reports 51B N-gram parameters, and Transformers
5.16.1 models the table as one very large `nn.Embedding`. Naively adding Qwen4
to Soup's streaming allowlist would therefore put the PLE table in one decoder
layer shard and require a layer-sized device buffer for the entire table. That
would defeat the bounded-memory property of layer streaming.

The PLE base table is frozen during LoRA and is accessed through sparse row
lookups. It is therefore a good candidate for read-only SSD-backed access: map
the safetensors data, read only the selected rows, and keep the table out of
both the resident base and the decoder-layer buffer pool.

Evidence from an Apple M4 Max with 128 GiB unified memory:

- the public `Jundot/Qwen3.8-Flash-Next-oQ4e-mtp` checkpoint is about 99 GiB;
- its affine PLE tensors occupy about 29.80 GiB across 128 shards;
- oMLX 0.6.4 already demonstrates the runtime shape of this optimization with
  a read-only `mmap` PLE row reader and a `qwen4_ple_ssd_offload` setting.

This is evidence that sparse SSD residency is viable, not evidence that Soup
can train the oMLX-specific oQ checkpoint directly.

## Proposed solution

Add a bounded Qwen4-Exp PLE path to Transformers layer streaming, exposed as:

```yaml
training:
  stream_layers: true
  stream_source: auto
  stream_ngram_source: disk  # auto / ram / disk
```

Proposed contract:

- initially accept the setting only for text `qwen4_exp_text` layer-streamed
  LoRA runs;
- locate the PLE N-gram tensor through the safetensors index and validate its
  rank, dtype, offsets, bounds, and source containment before mapping it;
- use a read-only mapping and gather only rows referenced by the current
  N-gram ids;
- exclude the PLE weight from the ordinary decoder-layer shard and buffer
  budget;
- keep the non-PLE PLE projection/normalization/conv weights in the ordinary
  streamed layer path;
- make `auto` choose between RAM and disk from measured table size and
  available RAM, while explicit `ram` / `disk` remain authoritative;
- close every mapping on normal completion and failure;
- fail closed on unsupported checkpoint layouts instead of materializing the
  table or silently falling back;
- admit `qwen4_exp_text` to the streaming allowlist only after resident versus
  streamed forward/backward parity is proven.

The source mapping must remain read-only. Training outputs, adapters, caches,
and checkpoints are still allowed to write normally; this setting is not a
claim that the whole training run is write-free.

## Acceptance evidence

- focused safetensors tests for header validation, row bounds, dtype handling,
  path containment, duplicate indices, cleanup, and no source mutation;
- exact row-gather parity against a resident embedding on synthetic shards;
- a tiny real `Qwen4ExpForCausalLM` resident-versus-SSD test proving finite and
  matching loss plus LoRA gradients through the PLE path;
- a mutation that restores whole-table layer buffering must fail the named
  regression;
- a real public-checkpoint smoke on the M4 Max once a Transformers-compatible
  checkpoint is available locally, with memory and throughput reported as
  observations rather than general capacity claims;
- full Soup test suite and `ruff check src/soup_cli/ tests/` clean.

## Alternatives considered

1. **Only add Qwen4 to `SUPPORTED_STREAM_ARCHS`.** Unsafe: the N-gram table
   becomes part of one enormous decoder layer and breaks the buffer invariant.
2. **Keep the whole table in CPU RAM.** This is the current Transformers-style
   fallback, but it consumes roughly the memory layer streaming is intended to
   save and is not viable on many target machines.
3. **Use generic `stream_source: disk` only.** Insufficient: Soup's disk tier
   still reads a whole layer into a device buffer. PLE needs sparse row-level
   access independent of the ordinary layer source tier.
4. **Target the installed oMLX oQ format first.** Rejected for the initial
   scope. Soup should establish the official Transformers/safetensors contract
   before considering an oMLX-specific affine layout.

## Scope boundaries

This issue does not claim:

- multimodal Qwen4 fine-tuning;
- direct training of oMLX oQ checkpoints;
- layer-streamed PEFT `target_parameters` for routed expert tensors (currently
  resident-only and suitable for a separate follow-up);
- MTP training or serving integration;
- a catalog recipe;
- performance parity with RAM or resident execution before measurement.

I am working on this.

Official references:

- https://huggingface.co/Qwen/Qwen3.8-Flash-Next
- https://github.com/huggingface/transformers/blob/v5.16.1/src/transformers/models/qwen4_exp/modeling_qwen4_exp.py
- #571
- #572
- #573
- #575

## Area

Training (SFT/DPO/GRPO)

**Comment by MakazhanAlpamys** [comment]
Addressed by `c97773b` (#603, by @Amix29), with a maintainer follow-up in `9c88efc`. **This issue stays open** on the production-checkpoint run.

What landed: Qwen4-Exp layer streaming with SSD-backed PLE n-gram embeddings, four fail-closed policies (mode / source / disk / RAM-fit), `config.json` folded into the oQ shard-cache fingerprint on both the sharder and planner sides, and oQ PLE sizing reported as packed `storage_nbytes` rather than the ~4x larger dense logical table.

The finding worth recording, because it is the one that took two rounds: the gates existed and nothing proved they were **reached**. Deleting both gate call sites gave a byte-identical `40 passed` — a fully green suite with the feature's fail-closed policies removed. Gate bodies are now pinned by behaviour tests and gate call sites by a recorder test that drives the real `_setup_streaming_transformers()`; every deletion combination fails it by name.

Also fixed on the way, and larger than a missing test: the RAM refusal keyed on `stream_source` rather than `stream_ngram_source`, so on the **default** `stream_source: auto` a 102 GB PLE table was loaded into RAM with no refusal.

Still open:

- [ ] a production-checkpoint run — `benchmarks/gate-qwen4-ple-m4-max.md` labels its stopped production step **not validated**, and `benchmarks/README.md` carries the matching partial-pass row
- [ ] the four numerical parity tests are skipped anywhere `transformers` is below the declared floor, so it is worth confirming CI installs `>=5.16.1` **and** that the installed build ships `qwen4_exp` — otherwise they skip in CI too and the parity claim is machine-checked nowhere

Separately filed: #636, on the `[train]` torch floor this PR raised being inconsistent with what `transformers>=5.16.1` already forces.
